harness/determined/experimental/inference/_torch_batch_process.py

```diff
@@ -0,0 +1,405 @@
+                default_output_uuid = metadata["default_output_uuid"]
+
+        output_uuid_with_rank = default_output_uuid + f"/rank_{rank}"
+
+        batch_processor_context = TorchBatchProcessorContext(core_context, output_uuid_with_rank)
+
+        per_batch_processor = batch_processor_cls(
+            context=batch_processor_context,
+        )
+
+        dataloader = DataLoader(
+            dataset=dataset, batch_size=batch_size, shuffle=False, **dataloader_kwargs
+        ).get_data_loader(repeat=False, skip=skip, num_replicas=total_worker, rank=rank)
+
+        # Create dummy searcher op to report progress to master
+        dummy_searcher_op = None
+        # Initialize dummy searcher for progress report
+        if rank == 0:
+            dummy_searcher_op = core.DummySearcherOperation(1, True)
+
+        dataloader_iterator = iter(dataloader)
+
+        # Enumerate over dataloader directly may cause some workers to iterate for 1 more time
+        # than others when drop_last = False. If those workers synchronize on the last batch_idx,
+        # they would hang forever as other workers never hit that last batch_idx.
+        # To avoid the issue, we calculate and take the ceiling of the iteration count to ensure
+        # all workers iterate for the same number of times.
+        max_batch = math.ceil(dataset_len / batch_size / total_worker)
+        iterate_length = _validate_iterate_length(iterate_length, max_batch)
+
+        last_checkpoint_idx = 0
+        batch_idx = skip
+        steps_completed = skip
+
+        """
+        (2) Run batch processing
+        """
+
+        for batch_idx in range(skip, iterate_length):
+            X = next(dataloader_iterator, None)
+            if X is not None:
+                per_batch_processor.process_batch(batch=X, batch_idx=batch_idx)
+            steps_completed = batch_idx + 1
+
+            # Checkpoint and check preemption
+            if (batch_idx + 1) % checkpoint_interval == 0:
+                logging.info(f"Completed steps:  {steps_completed} and checkpointing")
+                per_batch_processor.run_before_checkpoint()
+                _synchronize_and_checkpoint(
+                    core_context, steps_completed, rank, default_output_uuid
+                )
+                last_checkpoint_idx = batch_idx
+                # Report progress can only be done accurately with synchronization
+                core_context._tensorboard_manager.sync()
+                if rank == 0:
+                    _report_progress_to_master(
+                        dummy_searcher_op, batch_idx, total_worker, batch_size, dataset_len
+                    )
+                # Check preemption
+                if core_context.preempt.should_preempt():
+                    # Finish reducing metrics and report to not lose state before preempting
+                    _reduce_metrics(batch_processor_context, core_context, rank, steps_completed)
```

Comment:
If preemption is requested, we now just reduce metrics based on what we have right now and report. When trial is resumed, the metrics accumulation and reduction will start again.

Not sure if this is sufficient or we somehow need to persist the state of the metrics reducer.

Reply:
i think this is fine. it makes sense to me that we'd do a final dump of the metrics state before exiting. should we checkpoint before exiting though?

Reply:
We now only check preemption at the same interval as checkpointing, thus, a checkpoint consisting of most recent progress would have been created prior to us hitting this line.